### PR TITLE
SONAR-8467 Create WS api/project_analyses/search

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        .peek(addProject(dbSession, request))
+        .peek(checkPermission())
+        .peek(addAnalyses(dbSession, request))
+        .peek(addEvents(dbSession))
+        .map(SearchResults.Builder::build)
+        .collect(toOneElement());
+    }
+  }
+
+  private Consumer<SearchResults.Builder> addAnalyses(DbSession dbSession, SearchRequest request) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: MAJOR Remove this unused method parameter "request". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("search")
+      .setDescription("Search a project analyses and attached events.<br>" +
+        "Requires the following permission:" +
+        "<ul>" +
+        "  <li>'Browse' on the specified project</li>" +
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: I think you can remove the ul/li when there's only one permission

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      .setHandler(this);
+
+    action.addPagingParams(DEFAULT_PAGE_SIZE, 500);
+
+    action.createParam(PARAM_PROJECT)
+      .setDescription("Project key")
+      .setRequired(true)
+      .setExampleValue(KeyExamples.KEY_PROJECT_EXAMPLE_001);
+
+    action.createParam(PARAM_CATEGORY)
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: I've tried to use the category parameter (with OTHER or QUALITY_GATE), but events without category and events of type Version were not returned, events of type Quality Gate were still returned.

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchResults.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  private ListMultimap<String, EventDto> buildEvents(List<EventDto> events) {
+    return events.stream().collect(Collectors.index(EventDto::getAnalysisUuid));
+  }
+
+  static Builder builder(DbSession dbSession, SearchRequest request) {
+    return new Builder(dbSession, request);
+  }
+
+  static class Builder {
+    private final DbSession dbSession;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchResults.java]: dbSession is nerver used

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        .setPage(request.mandatoryParamAsInt(Param.PAGE))
+        .setPageSize(request.mandatoryParamAsInt(Param.PAGE_SIZE))
+        .build();
+    };
+  }
+
+  private SearchResults search(SearchRequest request) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      return Stream.of(SearchResults.builder(dbSession, request))
+        .peek(addProject(dbSession, request))
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: Interesting pattern, I think I would have go further by moving the addProject(), addAnalyses(), addEvents() to the SearchResults.Builder class

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: WDYM? I'm not sure I understand

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: How could I return event without category?

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java (hunk 5 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: similar to what I have done with the response, I understand. I thought about it, will give it another try

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchResults.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchResults.java]: refactoring not complete, thanks for the reminder

File: server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java]: as discussed, I update the description. All events of an analysis are returned

**Pull request merged**